org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java

```diff
@@ -0,0 +1,150 @@
+import org.eclipse.xtext.ui.generator.trace.ITraceForStorageProvider;
+
+import com.google.common.base.Preconditions;
+import com.google.inject.Inject;
+
+public class XbaseEditorInputRedirector {
+
+	private final static Logger LOG = Logger.getLogger(XbaseEditorInputRedirector.class);
+
+	@Inject
+	private ITraceForStorageProvider traceInformation;
+
+	@Inject
+	private LanguageInfo languageInfo;
+
+	@Inject
+	private FileExtensionProvider fileExtensionProvider;
+
+	public ITypeRoot getTypeRoot(IEditorInput editorInput) {
+		IJavaElement adapter = editorInput.getAdapter(IJavaElement.class);
+		return adapter instanceof ITypeRoot ? (ITypeRoot) adapter : null;
+	}
+
+	/**
+	 * @return the original source for an editor input that points to an Xtext resource copied to the output folder, the given input
+	 *         otherwise
+	 */
+	public IEditorInput findOriginalSourceForOuputFolderCopy(IEditorInput input) {
+		try {
+			IFile resource = ResourceUtil.getFile(input);
+			if (resource != null) {
+				// if the given resource is already pointing to a language file
+				if (fileExtensionProvider.isValid(resource.getFullPath().getFileExtension())) {
+					IJavaProject project = JavaCore.create(resource.getProject());
+					if (project.exists()) {
+
+						// Check if it's sitting in the output dir, look for original one
+						if (project.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+							IPath relative = resource.getFullPath().removeFirstSegments(project.getOutputLocation().segmentCount());
+
+							for (IPackageFragmentRoot root : project.getPackageFragmentRoots()) {
+								if (root.getKind() == IPackageFragmentRoot.K_SOURCE) {
+									IPath fullPath = root.getCorrespondingResource().getProjectRelativePath().append(relative);
+									IFile newFile = resource.getProject().getFile(fullPath);
+									if (newFile.exists()) {
+										return new FileEditorInput(newFile);
+									}
+								}
+							}
+						}
+
+						// Check if it's sitting in one of the output folders set on the source folders
+						for (IClasspathEntry classpathEntry : project.getRawClasspath()) {
+							if (classpathEntry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+								if (classpathEntry.getOutputLocation() != null
+										&& classpathEntry.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+									IPath relative = resource.getFullPath()
+											.removeFirstSegments(classpathEntry.getOutputLocation().segmentCount());
+									for (IPackageFragmentRoot root : project.findPackageFragmentRoots(classpathEntry)) {
+										IPath fullPath = root.getCorrespondingResource().getProjectRelativePath().append(relative);
+										IFile newFile = resource.getProject().getFile(fullPath);
+										if (newFile.exists()) {
+											return new FileEditorInput(newFile);
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		} catch (JavaModelException e) {
+			if (e.isDoesNotExist()) {
```

Comment:
It's the opposite of what I tried to suggest: Only log, ig it is not `e.isDoesNotExist()`